### PR TITLE
Magmalte rock 1.8

### PR DESCRIPTION
**Pull request by saltiyazan**
Builds on[ PR #1  ](https://github.com/canonical/magma-nms-magmalte-rock/pull/1) to fix the rock and to use magma 1.8

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      - name: Setup LXD
+        uses: whywaita/setup-lxd@v1
+        with:
+          lxd_version: latest/stable
+
+      - name: Install rockcraft
+        run: |
+          sudo snap install --classic --channel edge rockcraft
+      - name: Build ROCK
+        run: rockcraft pack --verbose
```
Comment (gruyaume) [inline .github/workflows/build.yaml]: Please use the `canonical/craft-actions/rockcraft-pack@main` workflow. You can see an example [here](https://github.com/canonical/magma-orc8r-controller-rock/blob/main/.github/workflows/build.yaml)

Comment (gruyaume) [reply .github/workflows/build.yaml]: overall you can very much re-use the workflows from the project linked above (also applicable to `push.yaml`)

File: README.md [hunk 2]
```diff
@@ -1 +1,14 @@
...
+> **Warning**: This rock is under construction
+
+Rock for Magma's orchestrator NMS service built using 
+[rockcraft](https://github.com/canonical/rockcraft). It provides a direct port of this 
+[Dockerfile](https://github.com/magma/magma/blob/v1.6/nms/app/packages/magmalte/Dockerfile).
+
+## Usage
+
+```bash
+docker pull ghcr.io/gruyaume/magma-orc8r-nms-magmalte:1.8.0
```
Comment (gruyaume) [inline README.md]:
```suggestion
docker pull ghcr.io/canonical/magma-orc8r-nms-magmalte:1.8.0
```

File: README.md [hunk 3]
```diff
@@ -1 +1,14 @@
...
+docker run -it ghcr.io/gruyaume/magma-orc8r-nms-magmalte:1.8.0
```
Comment (gruyaume) [inline README.md]:
```suggestion
docker run -it ghcr.io/canonical/magma-orc8r-nms-magmalte:1.8.0
```

File: README.md [hunk 4]
```diff
@@ -1 +1,14 @@
-# magma-nms-magmalte-rock
\ No newline at end of file
+# Magma Orchestrator NMS Rock
+
+> **Warning**: This rock is under construction
```
Comment (gruyaume) [inline README.md]: Remove this line

File: rockcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    source: https://busybox.net/downloads/busybox-1.34.1.tar.bz2
+    build-packages:
+      - make
+      - gcc
+    stage-packages:
+      - libc6
+    override-build: |
+      mkdir -p $CRAFT_PART_INSTALL/bin/
+      make defconfig
+      make
```
Comment (gruyaume) [inline rockcraft.yaml]: Isn't there a more `rockcraft` way to do this part? Meaning leveraging one of the plugins and avoiding `override-build`. I know I wrote it but it's not super beautiful

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: rockcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      yarn cache clean
+
+      # Build
+      cp -rf ${CRAFT_PART_SRC}/nms/* ${CRAFT_PART_INSTALL}/
+      cd ${CRAFT_PART_INSTALL}/
+      yarn run build
+
+    override-stage: |
+      set -x
+      curl -s https://nodejs.org/dist/v16.14.1/node-v16.14.1-linux-x64.tar.gz | tar xzf - -C ${CRAFT_PART_INSTALL}/ --no-same-owner --strip-components=1
```
Comment (gruyaume) [inline rockcraft.yaml]: Maybe we should have a stage just to install yarn. I don't know if it needs to be in this one. Or maybe not, I'm not sure

File: README.md [hunk 7]
```diff
@@ -1 +1,12 @@
-# magma-nms-magmalte-rock
\ No newline at end of file
+# Magma Orchestrator NMS Rock
+
+Rock for Magma's orchestrator NMS service built using 
+[rockcraft](https://github.com/canonical/rockcraft). It provides a direct port of this 
+[Dockerfile](https://github.com/magma/magma/blob/v1.6/nms/app/packages/magmalte/Dockerfile).
```
Comment (ghislainbourgeois) [inline README.md]: Should this point to 1.8?
```suggestion
[Dockerfile](https://github.com/magma/magma/blob/v1.8/nms/app/packages/magmalte/Dockerfile).
```

File: rockcraft.yaml (hunk 5 above)
Comment (ghislainbourgeois) [reply rockcraft.yaml]: I just looked at the documentation for the `make` plugin, and it basically says to use `override-build` for what we need to do: https://snapcraft.io/docs/make-plugin

I think in that case, using the plugin brings no value.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: .github/workflows/push.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
+name: Push (main)
+
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - "rockcraft.yaml"
+      - "files/**"
+      - ".github/workflows/**.yaml"
```
Comment (Gmerold) [inline .github/workflows/push.yaml]: Do we care to push out the new rock when workflows change?

File: .github/workflows/push.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        run: |
+          sudo snap install --devmode --channel edge skopeo
+
+      - name: Install yq
+        run: |
+          sudo snap install yq
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: magma-orc8r-nms-magmalte-rock
```
Comment (Gmerold) [inline .github/workflows/push.yaml]: Artifact's name, as defined in `build.yaml`, is `rock` not `magma-orc8r-nms-magmalte-rock.`

File: README.md [hunk 10]
```diff
@@ -1 +1,12 @@
-# magma-nms-magmalte-rock
\ No newline at end of file
+# Magma Orchestrator NMS Rock
+
+Rock for Magma's orchestrator NMS service built using
```
Comment (Gmerold) [inline README.md]:
```suggestion
Rock for Magma Orchestrator's NMS service built using 
```
?

File: .github/workflows/build.yaml (hunk 1 above)
Comment (saltiyazan) [reply .github/workflows/build.yaml]: Done

File: README.md (hunk 2 above)
Comment (saltiyazan) [reply README.md]: ✅

File: README.md (hunk 3 above)
Comment (saltiyazan) [reply README.md]: ✅

File: README.md (hunk 4 above)
Comment (saltiyazan) [reply README.md]: ✅

File: rockcraft.yaml (hunk 5 above)
Comment (saltiyazan) [reply rockcraft.yaml]: I would have tested with the [kbuild](https://snapcraft.io/docs/kbuild-plugin) plugin from Snapcraft, but Rockcraft doesn't have the same at the moment.

File: rockcraft.yaml (hunk 6 above)
Comment (saltiyazan) [reply rockcraft.yaml]: I don't exactly get what you mean.
But I realised that we don't need to have node installed twice, so I removed the snap and moved this line to the build part.

File: README.md (hunk 7 above)
Comment (saltiyazan) [reply README.md]: ✅

File: .github/workflows/push.yaml (hunk 8 above)
Comment (saltiyazan) [reply .github/workflows/push.yaml]: Fixed this and added new workflow files.

File: README.md (hunk 10 above)
Comment (saltiyazan) [reply README.md]: ✅

File: .github/workflows/push.yaml (hunk 9 above)
Comment (saltiyazan) [reply .github/workflows/push.yaml]: fixed

File: tests/integration/test_integration.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Integration tests for the nms-magmalte rock."""
+
+import docker  # type: ignore[import]
+import yaml
+import requests  # type: ignore[import]
```
Comment (Gmerold) [inline tests/integration/test_integration.py]: You could avoid this ignore by adding `types-requests` to the `tox.ini`.

File: tests/tox.ini [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+[testenv:fmt]
+description = Apply coding style standards to code
+deps =
+    black
+    isort
+commands =
+    isort {[vars]integration_test_path}
+    black {[vars]integration_test_path}
+
+[testenv:lint]
```
Comment (Gmerold) [inline tests/tox.ini:34]: Should we also add `lint` and `static` to the CI since it's already here?

Comment (saltiyazan) [reply tests/tox.ini:34]: It would cover the integration test only, should we add it?

Comment (Gmerold) [reply tests/tox.ini:34]: We can either add it as additional checks or just remove the relevant sections from this file.

**Comment by saltiyazan** [comment]
[Issue](https://github.com/canonical/craft-actions/issues/6) created on Craft-actions to address the failure in the ci.

File: tests/tox.ini (hunk 12 above)
Comment (saltiyazan) [reply tests/tox.ini:34]: Added the checks :)

File: tests/integration/test_integration.py (hunk 11 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py]: ✅

File: tests/integration/test_integration.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+import unittest
+from pathlib import Path
+from time import sleep
+
+import docker  # type: ignore[import]
+import requests
+import yaml
+
+#MAGMALTE_DOCKER_URL = "http://172.17.0.3"
```
Comment (gruyaume) [inline tests/integration/test_integration.py]: :cry:

Comment (saltiyazan) [reply tests/integration/test_integration.py]: debugging github runner in progress :)

Comment (saltiyazan) [reply tests/integration/test_integration.py]: it should be all good now :)

File: .github/workflows/lint-report.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+name: Lint report
+
+on:
+  workflow_call:
+
+jobs:
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v2
```
Comment (ghislainbourgeois) [inline .github/workflows/lint-report.yaml]: In another workflow we use v3, we should probably standardize.
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/static-analysis.yaml [hunk 15]
```diff
@@ -0,0 +1,14 @@
+name: Static analysis
+
+on:
+  workflow_call:
+
+jobs:
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v2
```
Comment (ghislainbourgeois) [inline .github/workflows/static-analysis.yaml]:
```suggestion
      - uses: actions/checkout@v3
```

File: rockcraft.yaml [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    build-packages:
+      - make
+      - gcc
+    stage-packages:
+      - libc6
+    override-build: |
+      mkdir -p $CRAFT_PART_INSTALL/bin/
+      make defconfig
+      make
+      mv busybox $CRAFT_PART_INSTALL/bin/
```
Comment (ghislainbourgeois) [inline rockcraft.yaml]: Just thought about this, but `busybox` is in the Ubuntu repos, can't we use that package directly? We would not have to maintain this whole block.

File: tests/integration/test_integration.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    sslmode=disable
+"""
+
+
+class TestNmsMagmalteRock(unittest.TestCase):
+    """Integration tests for the nms-magmalte rock."""
+
+    def setUp(self) -> None:
+        """Run containers to test."""
+        super().setUp()
```
Comment (ghislainbourgeois) [inline tests/integration/test_integration.py]: This line does nothing.
```suggestion
```

File: tests/integration/test_integration.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        self.network.connect(magmalte_container)
+
+    def test_given_nms_magmalte_container_is_running_when_http_get_then_hello_message_is_returned(  # noqa: E501
+        self,
+    ):
+        """Test to validate that the container is running correctly."""
+        sleep(10)
+        response = requests.get(
+            f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
+        )
```
Comment (ghislainbourgeois) [inline tests/integration/test_integration.py]:
```suggestion
        for _ in range(10):
            response = requests.get(
                f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
            )
            sleep(1)
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: .github/workflows/lint-report.yaml (hunk 14 above)
Comment (saltiyazan) [reply .github/workflows/lint-report.yaml]: ✅

File: .github/workflows/static-analysis.yaml (hunk 15 above)
Comment (saltiyazan) [reply .github/workflows/static-analysis.yaml]: ✅

File: rockcraft.yaml (hunk 16 above)
Comment (saltiyazan) [reply rockcraft.yaml]: Really good catch, thanks.
Done

File: tests/integration/test_integration.py (hunk 17 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py]: 👍

File: tests/integration/test_integration.py (hunk 18 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py]: Done.

File: tests/integration/test_integration.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+            },
+        )
+        self.network.connect(magmalte_container)
+
+    def test_given_nms_magmalte_container_is_running_when_http_get_then_hello_message_is_returned(  # noqa: E501
+        self,
+    ):
+        """Test to validate that the container is running correctly."""
+        url = f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
+        for _ in range(60):
```
Comment (ghislainbourgeois) [inline tests/integration/test_integration.py]: This loop might be a bit long, it will basically retry for a minute.

**Review by ghislainbourgeois — APPROVED** [review]
Small nitpick on my side around the retry loop in the test, otherwise LGTM

File: tests/integration/test_integration.py (hunk 19 above)
Comment (saltiyazan) [reply tests/integration/test_integration.py]: I increased the time just to be sure not to fail the test if the pod takes longer than expected to start and run the service.

Comment (saltiyazan) [reply tests/integration/test_integration.py]: At the end I decreased it to 30 :)

File: tests/tox.ini [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+[testenv:static]
+description = Run static analysis checks
+deps =
+    -r{toxinidir}/requirements.txt
+    mypy
+    types-PyYAML
+    pytest
+    pytest-operator
+    juju
```
Comment (sanchezfdezjavier) [inline tests/tox.ini:61]: Same here?

File: tests/tox.ini [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+skip_missing_interpreters = True
+envlist = integration
+
+[vars]
+integration_test_path = {toxinidir}/integration/
+
+[testenv]
+deps =
+    pytest
+    pytest-operator
```
Comment (sanchezfdezjavier) [inline tests/tox.ini]: Do we need this guy?

File: tests/tox.ini (hunk 20 above)
Comment (sanchezfdezjavier) [inline tests/tox.ini]: Same here.

File: tests/pyproject.toml [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+check_untyped_defs = true
+allow_redefinition = true
+
+# Ignore libraries that do not have type hint nor stubs
+[[tool.mypy.overrides]]
+module = ["ops.*", "kubernetes.*", "flatten_json.*", "git.*"]
+ignore_missing_imports = true
+
+[[tool.mypy.overrides]]
+module = ["charms.*"]
```
Comment (sanchezfdezjavier) [inline tests/pyproject.toml]: I think we're not using this guy.

File: tests/pyproject.toml [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+show_traceback = true
+show_error_codes = true
+namespace_packages = true
+explicit_package_bases = true
+check_untyped_defs = true
+allow_redefinition = true
+
+# Ignore libraries that do not have type hint nor stubs
+[[tool.mypy.overrides]]
+module = ["ops.*", "kubernetes.*", "flatten_json.*", "git.*"]
```
Comment (sanchezfdezjavier) [inline tests/pyproject.toml]: I think we're not using this guys.

**Review by gruyaume — APPROVED** [review]

**Review by Gmerold — APPROVED** [review]

File: tests/pyproject.toml (hunk 23 above)
Comment (saltiyazan) [reply tests/pyproject.toml]: ✅

File: tests/pyproject.toml (hunk 22 above)
Comment (saltiyazan) [reply tests/pyproject.toml]: ✅

File: tests/tox.ini (hunk 20 above)
Comment (saltiyazan) [reply tests/tox.ini]: ✅

File: tests/tox.ini (hunk 21 above)
Comment (saltiyazan) [reply tests/tox.ini]: ✅

File: tests/integration/test_integration.py [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+POSTGRES_USER = "username"
+POSTGRES_PASSWORD = "password"
+POSTGRES_DB = "magma"
+DATABASE_SOURCE = f"""
+    dbname=magma
+    user={POSTGRES_USER}
+    password={POSTGRES_PASSWORD}
+    host=postgres_container
+    sslmode=disable
+"""
```
Comment (sanchezfdezjavier) [inline tests/integration/test_integration.py]: Are we using this guy?

Comment (saltiyazan) [reply tests/integration/test_integration.py]: no, not anymore.
✅

**Review by sanchezfdezjavier — COMMENTED** [review]
LGTM

**Review by sanchezfdezjavier — APPROVED** [review]

**Pull request merged**